pallets/creditcoin/src/lib.rs

```diff
@@ -968,94 +969,56 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		#[transactional]
 		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(3,1))]
-		pub fn register_transfer(
+		pub fn register_funding_transfer(
 			origin: OriginFor<T>,
 			transfer_kind: TransferKind,
-			gain: ExternalAmount,
-			order_id: OrderId<T::BlockNumber, T::Hash>,
+			deal_order_id: DealOrderId<T::BlockNumber, T::Hash>,
 			blockchain_tx_id: ExternalTxId,
 		) -> DispatchResult {
 			let who = ensure_signed(origin)?;
 
-			let (from_id, to_id, mut amount) = match &order_id {
-				OrderId::Deal(deal_order_id) => {
-					let order = try_get_id!(DealOrders<T>, &deal_order_id, NonExistentDealOrder)?;
+			let order = try_get_id!(DealOrders<T>, &deal_order_id, NonExistentDealOrder)?;
 
-					if gain.is_zero() {
-						// transfer for initial loan
-						(order.lender_address_id, order.borrower_address_id, order.terms.amount)
-					} else {
-						// transfer to repay loan
-						(order.borrower_address_id, order.lender_address_id, order.terms.amount)
-					}
-				},
-				OrderId::Repayment(_) => {
-					ensure!(gain.is_zero(), Error::<T>::RepaymentOrderNonZeroGain);
-					return Err(Error::<T>::RepaymentOrderUnsupported.into());
-				},
-			};
-
-			let from = Self::get_address(&from_id)?;
-			let to = Self::get_address(&to_id)?;
-
-			ensure!(from.owner == who, Error::<T>::NotAddressOwner);
+			let (transfer_id, transfer) = Self::register_transfer_internal(
+				who,
+				order.lender_address_id,
+				order.borrower_address_id,
+				transfer_kind,
+				order.terms.amount,
+				OrderId::Deal(deal_order_id),
+				blockchain_tx_id,
+			)?;
+			Self::deposit_event(Event::<T>::TransferRegistered(transfer_id, transfer));
 
-			ensure!(from.blockchain == to.blockchain, Error::<T>::AddressPlatformMismatch);
+			Ok(())
+		}
 
-			ensure!(from.blockchain.supports(&transfer_kind), Error::<T>::UnsupportedTransferKind);
+		#[transactional]
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(3,1))]
+		pub fn register_repayment_transfer(
+			origin: OriginFor<T>,
+			transfer_kind: TransferKind,
+			gain: ExternalAmount,
```

Comment:
Should this be the repayment_amount? Seems more intuitive to me, otherwise the user has to do the math to see what the gain was, when repayment_amount is readily available. Nothing is ever done with gain other than adding it order.terms.amount to get repayament_amount.